Warn if Windows compatibility mode flags are detected on startup

smallketchup82 opened the pull request:
Closes #22092

## Description
This PR adds a new component, `CompatibilityModeChecker.cs` to `osu.Desktop/Windows`, which does as the name implies.

Windows writes to `HKCU\Software\Microsoft\Windows NT\CurrentVersion\AppCompatFlags\Layers` to track compatibility flags for applications. The key names in the registry folder coincide with the path to the executables. Following this logic, we can check each path in the registry folder to see if they contain `osu!.exe` which would likely mean a path to the lazer installation. If this is detected, the component gives the user a notification to warn them about using compatibility mode.

Lazer in general already refuses to run under compatibility mode (since it requires windows 8.1 or later) but the same can't be said for other values, such as fullscreen optimizations, dpi optimizations, etc. By using the registry, we can make sure that virtually every value that would be in the properties -> compatibility tab is turned off. Windows automatically deletes the keys if there are no compatibility options set, so we really only have to check that a key pointing to the lazer executable exists, and if so, warn the user.

## Concerns
I do have a bit of concerns with this implementation. I could have addressed the concerns below before making the PR, but for the sake of simplicity, I decided to leave them out pending discussion.
1. Checking for `osu!.exe` seems a bit hacky, it would probably be better to try and find matches to the runtime executable. Although I'm not sure if this would work properly since it looks like there's two executables, an entry `osu!.exe`, and another `osu!.exe` in the squirrel installation folder
	- This could also possibly trigger if there are compatibility options set for stable
	- It could be wise to wait until peppy finishes his [squirrel update](https://github.com/ppy/osu/issues/14824#issuecomment-1891532983), then come back to this
2. The notification strings should probably be made translatable, both in this component and in `ElevatedPrivilegesChecker.cs`
3. There also exists `HKLM\Software\Microsoft\Windows NT\CurrentVersion\AppCompatFlags\Layers` which similarly holds compatibility flag information. In my experience, this is rarely used and is only written to if the user presses "change settings for all users" when making the compatibility mode changes. Although I personally don't think that it's worth it since a user isn't likely to do this, we could also track this to be safe (grab the array of names from `HKLM`, combine it with the array from `HKCU`, then look through the final concatenated array instead)
4. Instead of simply notifying the user, we could outright disallow using any compatibility mode flags as peppy discussed in the issue thread. I personally don't think we need to go to that extreme, but it is an option
5. We can also turn off the flags for the user by deleting the registry key, but I don't believe we should do this since it'd be best to avoid touching the registry unless we need to

If the concerns above aren't pressing, then this implementation should be capable of easily checking to see if there are any compatibility flags set for the application

smallketchup82 committed:
Add check for compatibility mode on game startup

smallketchup82 committed:
Fix peppy copyright disclaimer

smallketchup82 committed:
Remove comments about compatibility mode from `Program.cs` since this PR will be resolving it

smallketchup82 commented on osu.Desktop/Security/CompatibilityModeChecker.cs:
```diff
@@ -0,0 +1,58 @@
+    /// <summary>
+    /// Checks if the game is running with windows compatibility optimizations which could cause issues. Displays a warning notification if so.
+    /// </summary>
+    public partial class CompatibilityModeChecker : Drawable
```
`ElevatedPrivilegesChecker.cs` seems to inherit `Component`, however `Component` seems to just be an alias for `Drawable`. Also, looking at the docs for `Component`, it looks to be implied that `Component` is deprecated, or will be deprecated. I decided to just use `Drawable` here as a result, but if it would be cleaner to use `Component`, it can be used

frenzibyte replied:
Not quite, `Component` will eventually be changed so that `Drawable` inherits from `Component`. Use `Component` here. Only use `Drawable` when this class is meant to show something visually (this one doesn't, it signals the notification overlay to show a notification, not display something by itself).

smallketchup82 committed:
Inherit from component instead of drawable

bdach commented on osu.Desktop/Program.cs:
```diff
@@ -52,10 +52,6 @@ public static void Main(string[] args)
                 // See https://www.mongodb.com/docs/realm/sdk/dotnet/compatibility/
                 if (windowsVersion.Major < 6 || (windowsVersion.Major == 6 && windowsVersion.Minor <= 2))
                 {
-                    // If users running in compatibility mode becomes more of a common thing, we may want to provide better guidance or even consider
-                    // disabling it ourselves.
-                    // We could also better detect compatibility mode if required:
-                    // https://stackoverflow.com/questions/10744651/how-i-can-detect-if-my-application-is-running-under-compatibility-mode#comment58183249_10744730
```
This comment can't just be removed. If compatibility options set are preventing the game from being launched *this code must detect it too and notify the user about it*. Because the notification added later *won't ever show if the game doesn't launch*.

Probably make `checkCompatibilityMode()` from `CompatibilityModeChecker` `public static` and call that here to handle that case.

smallketchup82 replied:
I think that we should rewrite the comment instead of keeping it. My idea is that once discussions are over, this PR will completely alleviate the issues we have with compatibility mode, therefore this comment is unnecessary (since we'd be replacing it with a description of our new implementation).

It would probably be good to hear from peppy if he would rather have lazer send a notification, or if he'd rather have lazer refuse to run with any compatibility mode flags set. We can get around to adapting this comment to our approach when we solidify which way we want to do this.

bdach replied:
> My idea is that once discussions are over, this PR will completely alleviate the issues we have with compatibility mode

No it won't if _compatibility mode prevents the game from running_.

The messagebox underneath this comment should be _checking_ if compatibility mode is enabled and only bringing it up if it detects it being enabled.

smallketchup82 replied:
I understand now. The code underneath the comment already disallows the game from running with compatibility mode. In this case, it would be wise to put the compatibility mode logic above the windows version checking logic. I'll fix this and add another comment with the diff to see if its satisfactory

smallketchup82 commented on osu.Desktop/Security/CompatibilityModeChecker.cs:
```diff
@@ -0,0 +1,58 @@
+using osu.Game.Overlays;
+using osu.Game.Overlays.Notifications;
+
+namespace osu.Desktop.Security
```
I figured that this should be in security since it sorta keeps the game in a "secure state" by doing these checks, but I guess a better way is to say that it keeps the game in a stable state by doing these checks?

I could probably make a new folder called "Stability" and put it in there.

bdach replied:
Just put it in `osu.Game/Windows`.

smallketchup82 commented on osu.Desktop/Security/CompatibilityModeChecker.cs:
```diff
@@ -0,0 +1,58 @@
+
+            using var layers = Registry.CurrentUser.OpenSubKey(@"Software\Microsoft\Windows NT\CurrentVersion\AppCompatFlags\Layers");
+
+            return layers != null && layers.GetValueNames().Any(name => name.EndsWith("osu!.exe", StringComparison.OrdinalIgnoreCase));
```
The hesitation I have with this is catching both the entry executable, and the runtime executable. Having compatibility flags on both wouldn't be favourable, and using `(Assembly.GetExecutingAssembly().Location` would really only capture the runtime executable.

smallketchup82 committed:
Move to osu.Desktop/Windows

smallketchup82 commented on osu.Desktop/Security/CompatibilityModeChecker.cs:
```diff
@@ -0,0 +1,58 @@
+        private bool checkCompatibilityMode()
+        {
+            if (!OperatingSystem.IsWindows())
+                return false;
```
Setting `[SupportedOSPlatform("windows")]` for the class, and removing [these lines](https://github.com/smallketchup82/osu/blob/be2f3f1bc0fe5de3fd5a6f01a83ed9954d7b8a69/osu.Desktop/Windows/CompatibilityModeChecker.cs#L33-L34), results in the following errors for [these lines](https://github.com/smallketchup82/osu/blob/be2f3f1bc0fe5de3fd5a6f01a83ed9954d7b8a69/osu.Desktop/OsuGameDesktop.cs#L148-L149):

`This call site is reachable on all platforms. 'CompatibilityModeChecker' is only supported on: 'windows'.`

bdach replied:
See

https://github.com/ppy/osu/blob/498948f17d4e2cd3dcc205502053402a2721777b/osu.Desktop/OsuGameDesktop.cs#L67

That `getStableInstallPathFromRegistry()` inside call is also annotated with that same attribute.

smallketchup82 committed:
Move operating system check to outside

smallketchup82 committed:
Show a different messagebox if compatibility mode is detected during windows version check

SupDos commented:
>but the same can't be said for other values, such as fullscreen optimizations

Hope I'm not being too annoying here, but is there any reason why you're also including the "Disable full-screen optimizations" checkbox to show this warning? This is what tells Windows (or well, suggests...) to allow Exclusive Fullscreen instead of forcing the game to run in borderless, and isn't something that would affect the game negatively.

smallketchup82 commented:
> > but the same can't be said for other values, such as fullscreen optimizations
> 
> Hope I'm not being too annoying here, but is there any reason why you're also including the "Disable full-screen optimizations" checkbox to show this warning? This is what tells Windows (or well, suggests...) to allow Exclusive Fullscreen instead of forcing the game to run in borderless, and isn't something that would affect the game negatively.

Remember this checkbox is to *disable* these optimizations. It is a setting meant for compatibility (such as if a program is having issues). Lazer will be able to use borderless and exclusive normally since fullscreen optimizations aren't enabled by default. With the setting checked on, windows disables the optimizations that it would otherwise runs for lazer (such as optimizations on overlays). Additionally, fullscreen optimizations on windows were created especially for games. Giving the user the ability to disable these adds another vector for things to go wrong. Best not to risk it at all.

This article should explain full screen optimizations in depth: https://devblogs.microsoft.com/directx/demystifying-full-screen-optimizations/

If there's evidence that lazer is inadvertently flipping this setting on, then I could program in an exclusion for that specific case. But at present it doesn't seem like lazer needs to disable full-screen optimizations, since there's really no [downside to having it on](https://devblogs.microsoft.com/directx/demystifying-full-screen-optimizations/#:~:text=we%20have%20extensive%20performance%20data%20that%20indicates%20that%20almost%20all%20users%20who%20use%20fullscreen%20optimizations%20have%20equal%20performance%20to%20full%20screen%20exclusive.).

smallketchup82 committed:
Check if compatibility mode is checked for the runtime executable instead of checking if it is flipped on for any executable titled "osu!.exe"

smallketchup82 committed:
Code quality

- Make checkCompatibilityMode static
- Fixing naming for CheckCompatibilityMode
- Use explicit type for exePath

bdach commented:
So I tested this diff and... I think trying to find out if compatibility mode is enabled it makes windows defender angry...

What I've done is I set up a mock release with `osu-deploy` on this branch and installed it. The path to the executable needs fixing (right now it's pointing to the dll and not the exe), but after fixing that... I've had windows defender quarantine the game executable _twice_ due to `Trojan:Script/Wacatac.B!ml`. Which means that either my machine and build toolchain is compromised and I need to start tearing everything to the ground right now, or (more likely, looking at google results) this is a bullcrap "machine" "learning" "detection" scaremonger. That also means that we probably can't ship it either because users will get scared out of their minds.

Before you ask, this doesn't happen if I do the same local release procedure on `master`, so it looks pretty evident that it's the registry read that defender isn't liking.

smoogipoo commented:
It should not get quarantined after the app is signed (part of our deploy process), but regardless that is pretty grim especially if it affects development.

I agree this is not an important enough issue to be addressed, though.

smallketchup82 commented:
> So I tested this diff and... I think trying to find out if compatibility mode is enabled it makes windows defender angry...
> 
> What I've done is I set up a mock release with `osu-deploy` on this branch and installed it. The path to the executable needs fixing (right now it's pointing to the dll and not the exe), but after fixing that... I've had windows defender quarantine the game executable _twice_ due to `Trojan:Script/Wacatac.B!ml`. Which means that either my machine and build toolchain is compromised and I need to start tearing everything to the ground right now, or (more likely, looking at google results) this is a bullcrap "machine" "learning" "detection" scaremonger. That also means that we probably can't ship it either because users will get scared out of their minds. I don't think this is an important enough change to warrant the headache to clear the false positive later.
> 
> Before you ask, this doesn't happen if I do the same local release procedure on `master`, so it looks pretty evident that it's the registry read that defender isn't liking.

I'll look into this myself and seeing if i can resolve it. If not I'll probably close this PR myself since this issue definitely doesn't warrent dealing with defender. Could you send the code that points to the exe instead of the dll so that I can test it on my machine and try to find a workaround?

bdach commented:
The relevant change is

```diff
diff --git a/osu.Desktop/Windows/CompatibilityModeChecker.cs b/osu.Desktop/Windows/CompatibilityModeChecker.cs
index 31f32441d7..69765ed319 100644
--- a/osu.Desktop/Windows/CompatibilityModeChecker.cs
+++ b/osu.Desktop/Windows/CompatibilityModeChecker.cs
@@ -2,6 +2,7 @@
 // See the LICENCE file in the repository root for full licence text.
 
 using System;
+using System.Diagnostics;
 using System.Linq;
 using System.Reflection;
 using System.Runtime.Versioning;
@@ -35,9 +36,7 @@ public static bool CheckCompatibilityMode()
         {
             using var layers = Registry.CurrentUser.OpenSubKey(@"Software\Microsoft\Windows NT\CurrentVersion\AppCompatFlags\Layers");
 
-            string exePath = Assembly.GetExecutingAssembly().Location;
-
-            return layers != null && layers.GetValueNames().Any(name => name.Equals(exePath, StringComparison.OrdinalIgnoreCase));
+            return layers != null && layers.GetValueNames().Any(name => name.Equals(Environment.ProcessPath, StringComparison.OrdinalIgnoreCase));
         }
 
         private partial class CompatibilityModeNotification : SimpleNotification

```

I believe.

smallketchup82 commented:
Reading [this](https://stackoverflow.com/a/70314433/14688730), the issue might be with the exe location code rather than the registry?

smallketchup82 commented:
@bdach Quick question, if you still have the exe that's causing issues, could you send it to me (either via discord or a link)? I'd like to do malware analysis on it to figure out what fingerprints the ml is detecting in particular.

Apart from a conveniently timed notification from windows defender to do a security scan while starting lazer, I can't seem to reproduce this. Though I haven't yet tried it with osu-deploy (the lack of docs makes it a pain to figure out) though I'm working on trying to reproduce with osu-deploy.

bdach commented:
> Though I haven't yet tried it with osu-deploy (the lack of docs makes it a pain to figure out) though I'm working on trying to reproduce with osu-deploy.

Please do try that first. There should be no docs required. You just run the thing on windows. So long as `osu` and `osu-deploy` are in one common dir it should just work :tm: .

Although if you're going to try that then probably it's best that you apply

```diff
diff --git a/osu.Game/OsuGameBase.cs b/osu.Game/OsuGameBase.cs
index 81e3d8bed8..edd8246d73 100644
--- a/osu.Game/OsuGameBase.cs
+++ b/osu.Game/OsuGameBase.cs
@@ -98,7 +98,7 @@ public partial class OsuGameBase : Framework.Game, ICanAcceptFiles, IBeatSyncPro
         /// </summary>
         private const double global_track_volume_adjust = 0.8;
 
-        public virtual bool UseDevelopmentServer => DebugUtils.IsDebugBuild;
+        public virtual bool UseDevelopmentServer => true;
 
         public virtual EndpointConfiguration CreateEndpoints() =>
             UseDevelopmentServer ? new DevelopmentEndpointConfiguration() : new ProductionEndpointConfiguration();
@@ -110,7 +110,7 @@ public partial class OsuGameBase : Framework.Game, ICanAcceptFiles, IBeatSyncPro
         /// </summary>
         public string VersionHash { get; private set; }
 
-        public bool IsDeployedBuild => AssemblyVersion.Major > 0;
+        public bool IsDeployedBuild => false;
 
         internal const string BUILD_SUFFIX = "lazer";
 

```

so that you can use dev credentials and dev game instance for safer testing.

smallketchup82 commented:
Got it, will definitely try

smallketchup82 commented:
@bdach Using the exact patches and also osu-deploy guide which you sent, and simulating a mock release using osu-deploy, I cannot reproduce the windows defender issue.

I've also uploaded my output exe to virustotal for a 2nd opinion, it has come back entirely positive.
https://www.virustotal.com/gui/file-analysis/ZWIzODI5ZWY3M2FkMWMwOGEyZGJhMTJjMGExNDZkZjA6MTcxMTEyNjc4OA==

bdach commented:
I cannot reproduce it either now. But it did happen, I am not crazy.

![image](https://github.com/ppy/osu/assets/20418176/ee5d12c3-f120-4561-9f7a-038e9d7e005f)

It might be because windows nudged their "ai" "detection" bullshit.

![image](https://github.com/ppy/osu/assets/20418176/e8cb2485-1a26-4a73-85a5-ef30690ddbb3)

I don't even know where to begin with this now. But I'm still reluctant to merge this after that experience. I'm not sure it's even providing that much benefit to anyone.

smallketchup82 commented:
> I cannot reproduce it either now. But it did happen, I am not crazy.
It was probably resolved by a windows defender definition update in that case.

> But I'm still reluctant to merge this after that experience.
I don't think that the scare warrants not merging. Sure this isn't that important of a change, but it does help defend against end users messing up their configuration then coming here complaining. It also seemed to be an edge case experienced by only you, since my builds (despite not using osu-deploy) also didn't return anything after checking for exclusions, doing a local defender scan, and doing a virustotal scan.

I do agree that we should block the pr for a month then revisit after we can be certain that a number of users have updated their definitions. Then we can also have someone like peppy, frenzi, or smoogi also try building to see what result they get.

peppy commented:
Shall we apply the filename fix and get this in?

As mentioned, windows defender pings aren't worth worrying about – that's for microsoft to fix (unless we're doing something *really* weird).

bdach commented:
If you're not worried then I guess :shrug:

smallketchup82 commented:
I'll commit the filename patch

But I would like to ask, wouldn't it be wise to log the value of the registry key after having checked for it? I feel as though if this is blocking people from launching the game, it would be wise to include *what* is blocking the game from launching in the debug logs. If someone comes here complaining about it, we can easily tell them what to turn off by looking at their debug logs instead of having to teach them to navigate the registry editor and to find the offending option.

Should I make that change or are we fine?

bdach commented:
I would like to think the only option that would block the game from launching is the "pretend to be windows 8 or earlier" option. But I guess no harm in enabling.

Susko3 commented on osu.Desktop/Windows/CompatibilityModeChecker.cs:
```diff
@@ -0,0 +1,60 @@
+        {
+            using var layers = Registry.CurrentUser.OpenSubKey(@"Software\Microsoft\Windows NT\CurrentVersion\AppCompatFlags\Layers");
+
+            string exePath = Assembly.GetExecutingAssembly().Location;
```
This doesn't actually work. Contrary to the variable name, this returns the full path of `osu!.dll`, but the registry has the path of `osu!.exe`.

There already exists a perfectly good way to get the exe path in `WindowsAssociationManager`. You should move that to a common windows helper class and use it instead.

https://github.com/ppy/osu/blob/3df32638c2235029625d929c41ce6237c646e75c/osu.Desktop/Windows/WindowsAssociationManager.cs#L32

---

It's also important to note that enabling compatibility mode on the osu!(lazer) start menu shortcut enables it only on the squirrel shim `osu!.exe` and not `osu!.exe` that is tested here. So I'm not certain that this will actually work on deployed builds.

bdach replied:
[maybe scroll up a few](https://github.com/ppy/osu/pull/27654#issuecomment-2013969557)

smallketchup82 replied:
The logic for locating the executable was incorrect in the initial PR. @bdach gave a patch earlier, replacing `Assembly.GetExecutingAssembly().Location;` with `Environment.ProcessPath` which fixes this issue. I'm going to commit this patch along with some other final touches.

Unless `Environment.ProcessPath` doesn't work (which it should, since I've tested it on deployed builds), we should be fine with using that.

smallketchup82 commented:
@bdach Along with some of my final touches, are we sure that we don't want to make the notification string translatable (one of my original concerns listed)?

bdach commented:
> notification string translatable

not sure why that requires pinging me specifically but whatever

the in-game text you can make translatable i guess. you're gonna have more trouble with the text that shows when the game refuses to launch because at that point none of the framework localisation machinery is active yet. so i'd suggest not bothering with that

smallketchup82 committed:
Final touches

- Add docstrings to the functions
- Add a new function to log the compatibility flags
- Add patch for locating the actual exe

smallketchup82 committed:
Move flags log within if statement clause

In my previous commit I figured that I should put it outside of the if
statement so that it logs the compatibility mode flags regardless (in
case the other messagebox is shown for some reason and compatibility
mode is enabled). But I figured that it's highly likely this would never
happen, so might as well just keep it in the if statement.

smallketchup82 committed:
Make the notification text for the compatibility mode checker translatable

Will probably make a small PR in the future to do the same for the admin
mode checker

smallketchup82 commented:
Should be all good for merge

Susko3 commented on osu.Desktop/Windows/CompatibilityModeChecker.cs:
```diff
@@ -0,0 +1,76 @@
+            using var layers = Registry.CurrentUser.OpenSubKey(@"Software\Microsoft\Windows NT\CurrentVersion\AppCompatFlags\Layers");
+
+            if (layers?.GetValue(Environment.ProcessPath) is string flags)
+                Logger.Log($"Compatibility flags for {Environment.ProcessPath}: {flags}", LoggingTarget.Information);
```
```suggestion
                Logger.Log($"Compatibility flags for {Environment.ProcessPath}: {flags}");
```

Maybe don't log to [the void](https://github.com/ppy/osu-framework/blob/3f59b971be9563624731a15cfed3fd01b2783a2d/osu.Framework/Logging/Logger.cs#L556-L559) if you intend for this to be visible in bug report log exports.

Susko3 commented on osu.Desktop/Program.cs:
```diff
@@ -52,15 +52,25 @@ public static void Main(string[] args)
+                            "osu! is running in compatibility mode",
+                            "osu! is running in compatibility mode. This may cause issues with the game. Please ensure osu! is not set to run in compatibility mode.", IntPtr.Zero);
+
+                        CompatibilityModeChecker.LogCompatibilityFlags();
```
This won't actually log anything to a logfile as the logger is not yet initialised.

smallketchup82 replied:
Not sure what could be done in this situation.. Logging to console wouldn't be very useful in this case. I could refactor the function to return the flags instead of returning void, then include it in the messagebox. What do you think about that approach?

Susko3 reviewed (changes requested):
Please test your changes ¯\\\_(ツ)_/¯

smallketchup82 committed:
Don't log to void

smallketchup82 committed:
Forgot to add the return description to the function

smallketchup82 committed:
Fix compatibility flags debug info not being written to logfiles due to uninitialized logger

This makes a new function, `GetCompatibilityFlags` and includes that
within the messagebox. This would probably be a better approach for
diagnosing something like this in general (instead of having to flip
through a user's debug logs, you simply view a screenshot of the
messagebox)
